index.js

```diff
@@ -1,251 +1,300 @@
+    if (payload.type === 'unsubscribed')
+      return console.log(`Unsubscribed from ${payload.market}`)
+    else if (payload.type === 'update' || payload.type === 'partial') {
+      const id = this.toId(payload.market, payload.channel)
+      this.emit(id, payload.data, e.data)
+    } else {
+      console.log(new Date(), '[FTX] unhandled WS event', payload)
     }
   }
 
   toId(market, channel) {
-    if(!market) {
-      return channel;
+    if (!market) {
+      return channel
     }
 
-    return market + '::' + channel;
+    return market + '::' + channel
   }
 
   sendMessage = async (payload) => {
-    if(!this.connected) {
-      if(!this.reconnecting) {
-        throw new Error('[ftx ws] Not connected.');
+    if (!this.connected) {
+      if (!this.reconnecting) {
+        throw new Error('[ftx ws] Not connected.')
       }
 
-      await this.afterReconnect;
+      await this.afterReconnect
     }
 
-    this.ws.send(JSON.stringify(payload));
+    this.ws.send(JSON.stringify(payload))
   }
 
   subscribe = async (channel, market = undefined) => {
+    const id = this.toId(market, channel)
 
-    const id = this.toId(market, channel);
-
-    if(!this.connected) {
-      if(!this.reconnecting) {
-        throw new Error('[ftx ws] Not connected.');
+    if (!this.connected) {
+      if (!this.reconnecting) {
+        throw new Error('[ftx ws] Not connected.')
       }
 
-      await this.afterReconnect;
+      await this.afterReconnect
     }
 
-    if(this.subscriptions.map(s => s.id).includes(id)) {
-      return console.error(new Date, 'refusing to channel subscribe twice', market, channel);
+    if (this.subscriptions.map((s) => s.id).includes(id)) {
+      return console.error(
+        new Date(),
+        'refusing to channel subscribe twice',
+        market,
+        channel
+      )
     }
 
     const sub = {
       id,
       channel,
       market,
       doneHook: false,
-      done: false
+      done: false,
     }
 
-    this.subscriptions.push(sub);
-    this._subscribe(sub);
+    this.subscriptions.push(sub)
+    this._subscribe(sub)
 
-    return sub.done;
+    return sub.done
   }
 
   _subscribe(sub) {
-
-    sub.done = new Promise(r => sub.doneHook = r);
+    sub.done = new Promise((r) => (sub.doneHook = r))
 
     const message = {
       op: 'subscribe',
       market: sub.market,
-      channel: sub.channel
+      channel: sub.channel,
     }
 
-    this.sendMessage(message);
+    this.sendMessage(message)
   }
-}
 
+  unsubscribe = async (channel, market = undefined) => {
+    const id = this.toId(market, channel)
+
+    if (!this.connected) {
+      if (!this.reconnecting) {
+        throw new Error('[ftx ws] Not connected.')
+      }
+
+      await this.afterReconnect
+    }
+
+    const sub = {
+      id,
+      channel,
+      market,
+      doneHook: false,
+      done: false,
+    }
+    this.subscriptions = this.subscriptions.map((s) => s.id !== sub.id)
```

Comment:
you mean filter? this will return an array of bools

Reply:
Ah yes I did mean filter. will update